app/queries/keywords_query.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class KeywordsQuery
+  attr_reader :keywords
```

Comment:
I think it should be private, as we are accessing the word like  `KeywordsQuery.new(User.new).keywords`

```ruby
...
private

attr_reader :keywords
...
```

Reply:
But then we do not need the attribute at all, no? :) 
(I've added the attribute as it was part of an example of a QueryObject pattern -- but indeed, if we do not need it, there is no point to have it).
I've updated the code to remove the attr, let me know if it's ok @junan  :)

Reply:
@malparty `attr_reader` actually  generates  a `getter`  method behind the scene. When we read data, we dont directly call the instance variable(ex: @keywords), rather we retrieve the data with a getter method (ex: `keywords` method which generated by `attr_reader :keywords`).

For example this class
```Ruby
class SomeClass
  # Define getter method in public if we want to access a, b, c outside of the class
  attr_reader :a, :b, :c
  
  def initialize(a, b, c)
    @a = a
    @b = b
    @c = c
  end

  def call
    # access with a/b/c using getter method instead of accessing @a/@b/@c directly
  end
  
  private
  
  # Define getter method in private if we dont want to access a, b, c outside of the class
  attr_reader :a, :b, :c

  def some_private_method
    # access with a/b/c using getter method instead of accessing @a/@b/@c directly
  end
end
```

So your class will be like this

```Ruby
class KeywordsQuery
  def initialize(user)
    @keywords = user.keywords
  end

  def call
    order_by_name
  end

  private

  attr_reader: keywords
  
  def order_by_name
    keywords.order(:name)
  end
end
```

Reply:
Thanks for the example, though I'm not sure to understand the 'why'. 

So you mean any single private instance variable should never be accessed directly, but always via accessors? 
Is it to keep the flexibility of overriding the `getter` method later? Or any other reason? 

I'm asking to be sure I understand the why -- back in C# I used to utilize getter/setters solely for public instance variables (called "Properties"). Here the why was specific to maintain a stable contract (API) so that consumers of the class won't have to change when the variable does (as getter will maintain a bridge). But for private variables, I'm not sure to see exactly what we want to tackle~ 

Though I will refactor now as you mentionned ;-)

Thanks @junan :)

Reply:
@malparty Theoretically we can access the instance variable directly inside the class, but we generally [access](https://github.com/nimblehq/the1-web/blob/development/engines/the1_auth/app/forms/the1_auth/token_form.rb) them with accessors. I see the following benefit accessing value with accessor methods rather than directly accessing them with instance variable  
- Defining `attr_reader` clearly tells us that those attributes are only for reading(not writing), so more clarity
- If we need to access those variables outside of class, we cant access them directly with the instance variable. We need to define the accessor. So It is better to use the accessor method always to access data, so more consistency
- Every time we access the variable value inside the class, we type one less character(`@`), so less typing

Reply:
Thanks, very clear and I did not think of the attr-reader only situation!!! Cool! :)